### ISSUE TITLE
`torch.nn.functional.conv2d` (CPU) is very slow on a specific trained weight

### ISSUE THREAD

**Issue by pgmmpk** [pytorch/pytorch#19651]
## 🐛 Bug?

`torch.nn.functional.conv2d` (CPU) is very slow on a specific trained weight: 50x slower than using a random weight. Same shapes, same input, but:
```
Fast conv2d (random weights) Time elapsed:  0.08310413360595703
Slow conv2d (trained weights) Time elapsed:  3.3756301403045654
```

## To Reproduce

Use the gist: https://gist.github.com/pgmmpk/b64901e3bd77ed58c00be83bd170d982

what it does:
1. Downloads trained weights (18Mb)
2. creates input tensor (normal random)
3. computes convolution with trained weights - takes 3+ secs!
5. computes convolution with randomly initialized weights - takes 0.08 sec

## Expected behavior

Expect similar execution time: conv2d should take a fixed number of FLOPs

## Environment

```
Collecting environment information...
PyTorch version: 1.0.1.post2
Is debug build: No
CUDA used to build PyTorch: None

OS: Mac OSX 10.14.4
GCC version: Could not collect
CMake version: Could not collect

Python version: 3.7
Is CUDA available: No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

Versions of relevant libraries:
[pip3] numpy==1.16.3
[pip3] torch==1.0.1.post2
[pip3] torchvision==0.2.2.post3
[conda] Could not collect
```

## Additional context

Same behavior in 0.4.1

**Comment by vadimkantorov** [comment]
maybe you're hitting a problem with denormalized weights? you may try setting `torch.set_flush_denormal(True)` and see if that helps

**Comment by soumith** [comment]
this is indeed because some of the weights are in denormal range, and Intel / ARM CPUs take a very slow path for floating point math in such cases. The workaround that @vadimkantorov posted should work for you.

You can read more about it here: https://stackoverflow.com/questions/36781881/why-denormalized-floats-are-so-much-slower-than-other-floats-from-hardware-arch

**Comment by pgmmpk** [comment]
alas, changes nothing

**Comment by pgmmpk** [comment]
@soumith @vadimkantorov suggested workaround **did not work**. Please advise

**Comment by pgmmpk** [comment]
@soumith @vadimkantorov I confirm that problematic weight contains a lot of subnormal numbers. So this is the cause of slowness.

However, suggested workaround did not help.

```
print( (weight.abs() < 1.e-32).float().sum().item() )
print( (weight_random.abs() < 1.e-32).float().sum().item() )
>>
4533372.0
0.0
```

When I clamp problematic weights to 0, slowness goes away.

**Comment by albanD** [comment]
Hi,

First please avoid cross posting on the forum and here as we end up answering twice to the same questions and that's wasting people's time.

Second, after adding `torch.set_flush_denormal(True)` to your script just after the imports, I see the following output on my machine:
```
(releaseenv3.6) alban@obelix:~$ python workspace/dev-pytorch/test/tmp.py 
Trained weight
	shape: torch.Size([512, 1024, 3, 3])
	dtype: torch.float32
	any nans?: 0
	min: -0.07247687131166458
	max: 0.0953093096613884
	mean: -4.6253231289483665e-07
	std: 0.0008879730594344437
Good (random) weight
	shape: torch.Size([512, 1024, 3, 3])
	dtype: torch.float32
	any nans?: 0
	min: -0.0048497761599719524
	max: 0.004831577185541391
	mean: -3.502289587231644e-07
	std: 0.001000691088847816
Input vector
	shape: torch.Size([1, 1024, 32, 32])
	dtype: torch.float32
	any nans?: 0
	min: -5.017491817474365
	max: 4.822595119476318
	mean: -0.0013921058271080256
	std: 1.0015257596969604
Fast conv2d (random weights) Time elapsed:  0.09646725654602051
Slow convd (trained weights) Time elapsed:  0.09580278396606445
```

It does not work for you?

**Comment by pgmmpk** [comment]
Thank you for looking into it. And sorry for cross-posting.

No, the suggested workaround does not work for me.

However, I can confirm that if I clamp small weight values to zero, problem goes away.

**Comment by albanD** [comment]
Can you run the following modification of your code?

```python
'''
Demostrates what seem to be a performance bug (or puzzle) in PyTorch:
conv2d operation on trained weights is 20-50x slower than using random weights
'''

import contextlib
import torch
from torch import nn
import time
import contextlib
import urllib.request
import os

if not torch.set_flush_denormal(True):
    print("Unable to set flush denormal")
    print("Pytorch compiled without advanced CPU")
    print("at: https://github.com/pytorch/pytorch/blob/84b275b70f73d5fd311f62614bccc405f3d5bfa3/aten/src/ATen/cpu/FlushDenormal.cpp#L13")

def download_as(url, filename):
    if os.path.exists(filename):
        return  # do not download twice
    print('Downloading', url, 'as', filename)
    data = urllib.request.urlopen(url).read()
    with open(filename, 'wb') as f:
        f.write(data)

@contextlib.contextmanager
def timeit(message=''):
    start = time.time()
    try:
        yield
    finally:
        print(message, 'Time elapsed: ', (time.time() - start))

def printit(header, x):
    print(header)
    print('\tshape:', x.shape)
    print('\tdtype:', x.dtype)
    print('\tany nans?:', torch.isnan(x).any().item())
    print('\tmin:', x.min().item())
    print('\tmax:', x.max().item())
    print('\tmean:', x.mean().item())
    print('\tstd:', x.std().item())


# 18Mb file with trained weights:
download_as(
    'https://drive.google.com/uc?export=download&id=15Z6ACJKRep9Wjl_heyG1WS34Vsb_IUZn',
    'trained_weights_numpy.pkl'
)
with open('trained_weights_numpy.pkl', 'rb') as f:
    import pickle
    z = pickle.load(f)
weight = torch.tensor(z)
printit('Trained weight', weight)

# random convolution weights
weight_random = torch.zeros(512, 1024, 3, 3)
weight_random.normal_(std=0.001)
printit('Good (random) weight', weight_random)

# random input vector
x = torch.zeros(1, 1024, 32, 32)
x.normal_()
printit('Input vector', x)

with timeit('Fast conv2d (random weights)'):
    torch.nn.functional.conv2d(x, weight_random)

with timeit('Slow convd (trained weights)'):
    torch.nn.functional.conv2d(x, weight)
```

In particular, the flush denormal method can fail if some cpu features were not available at compile time.

**Comment by pgmmpk** [comment]
`torch.set_flush_denormal(True)` returns `True` here. Running your version of the script still gives me slow behavior:

```
...
Fast conv2d (random weights) Time elapsed:  0.08171486854553223
Slow convd (trained weights) Time elapsed:  3.100939989089966
```

**Comment by albanD** [comment]
I can't reproduce that with the 3 machines with intel processors I have around (i7-5930K, i7-6850K, i7-7900X) some with source install / some with pip wheel install.

Could you give more informations about your CPU and how you installed pytorch please?

**Comment by pgmmpk** [comment]
MacBook Pro model A1398. 

PyTorch installed into virtual environment with `pip install torch torchvision`.

Trying to figure out how to get the processor model... Does this help?
![image](https://user-images.githubusercontent.com/569458/56654846-1adb0b80-665f-11e9-9835-53a69d1c2d0d.png)

**Comment by pgmmpk** [comment]
Yes, workaround does work on a (newer) linux box.

Problem seem to be with the older MacBook only.

I would declare the problem solved, since MacBook is a development machine, not running production inference.

**Comment by albanD** [comment]
Hi,

I confirm that this happens with a colleague's macbook here as well but on no linux.
@soumith do you think someone should look into this?

**Comment by soumith** [comment]
I think we should look into this. I opened a new issue here: https://github.com/pytorch/pytorch/issues/19672

**Comment by willbradshaw** [comment]
Hi,

I'm also seeing this issue, on a fairly new Linux machine (Dell XPS13 running Ubuntu 18.04.3 LTS, with a Intel® Core™ i7-8565U CPU @ 1.80GHz × 8 CPU). Here's what I get when I run the above code:

```
Trained weight
        shape: torch.Size([512, 1024, 3, 3])
        dtype: torch.float32
        any nans?: False
        min: -0.07247687131166458
        max: 0.0953093096613884
        mean: -4.6253231289483665e-07
        std: 0.0008879730594344437
Good (random) weight
        shape: torch.Size([512, 1024, 3, 3])
        dtype: torch.float32
        any nans?: False
        min: -0.005613739602267742
        max: 0.0049783275462687016
        mean: -1.2575401342473924e-07
        std: 0.001000270596705377
Input vector
        shape: torch.Size([1, 1024, 32, 32])
        dtype: torch.float32
        any nans?: False
        min: -4.707863807678223
        max: 4.841742992401123
        mean: 0.00109342101495713
        std: 1.0001064538955688
Fast conv2d (random weights) Time elapsed:  0.022645950317382812
Slow convd (trained weights) Time elapsed:  4.075138330459595
```

**Comment by AtomicCactus** [comment]
This problem seems to have been fixed for me on the latest Torch version (1.13.0). Before the update one of two similar Linux machines was running into this issue.

**Intel(R) Core(TM) i7-1087SH** - older CPU, did not have this problem
**12th Gen Intel(R) Core(TM) i7-12800H** - newer CPU had this problem

As other posters mentioned, flush denormals did not help on the newer CPU, however installing latest Torch fixed something and now they're roughly on par with each other.
